tests/test_genomes.py

```diff
@@ -72,3 +73,222 @@ def test_add_genomic_element_type(self):
+                    proportions=[0.6, 0.4],
+                    mutation_types=bad_mut_types,
+                )
+        for bad_sums in [[-0.4, 0.5], [0.6, 0.8], [139487135987, 0.0], [0.2, 0.3]]:
+            with pytest.raises(ValueError):
+                _ = stdpopsim.DFE(
+                    id="abc",
+                    description="test test",
+                    long_description="test test test test",
+                    proportions=bad_sums,
+                    mutation_types=[m1, m2],
+                )
+
+
+class DFEModelTestMixin:
+    """
+    Mixin for testing specific models. Subclasses should extend
+    this class and define the self.model (as the model instance).
+    """
+
+    # To be defined in subclasses.
+    model = None
+
+    def test_debug_runs(self):
+        dbg = self.model.model.debug()
+        dfe_model = self.dfe
+        assert dfe_model
+        assert dbg.num_epochs > 0
+
+    @pytest.mark.filterwarnings("ignore:.*Zigzag_1S14.*:UserWarning")
+    def test_simulation_runs(self):
+        # With a recombination_map of None, we simulate a coalescent without
+        # recombination in msprime, with madeup mutation rate.
+        contig = stdpopsim.Contig.basic_contig(
+            length=5000, mutation_rate=self.model.mutation_rate
+        )
+        contig.clear_genomic_mutation_types()
+        contig.mutation_rate = 2.35e-8  # should I define one or use the model one?
+        # adding dfe to the contig
+        contig.add_DFE(
+            intervals=np.array([[0, 5000]]),
+            DFE=self.dfe,
+        )
+
+        # Generate vector with 2 samples for each pop with sampling enabled
+        sample_count = []
+        for p in self.model.populations:
+            if p.allow_samples:
+                sample_count.append(2)
+            else:
+                sample_count.append(0)
+        samples = self.model.get_samples(*sample_count)
+        engine = stdpopsim.get_engine("slim")
+        ts = engine.simulate(
+            self.model, contig, samples, slim_scaling_factor=10, slim_burn_in=10
+        )
+
+        mut_info = {}
+        for mut in ts.mutations():
+            for j, md in zip(
+                mut.derived_state.split(","), mut.metadata["mutation_list"]
+            ):
+                if j not in mut_info:
+                    mut_info[int(j)] = md
+
+        num_nonneutral = sum([mut_info[j]["selection_coeff"] != 0.0 for j in mut_info])
+        assert ts.num_populations == self.model.num_populations
+        assert len(mut_info.keys()) > 0  # number of mutations
+        assert num_nonneutral > 0  # nonneutral mutations
+
+
+class CatalogDFEModelTestMixin(DFEModelTestMixin):
+    """
+    Mixin for demographic models in the catalog.
+    """
+
+    def test_id_valid(self):
+        assert utils.is_valid_dfe_model_id(self.dfe.id)
+
+
+qc_test_classes = []
+for species in stdpopsim.all_species():
+    for model in species.demographic_models:
+        for dfe in species.dfes:
```

Comment:
ah, I see! Nice work figuring this out. But, I don't think we need to test all combinations of DFE and model. I don't think we need to even test DFEs with non-generic models at all - they're pretty much completely separated?

Reply:
Hm - and, this won't test DFEs at all in species without demographic models. I'm replacing this with a generic model.